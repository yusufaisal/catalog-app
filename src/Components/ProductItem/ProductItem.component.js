import * as React from 'react';
import { View, Image, Text } from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';

import styles from './ProductItem.component.styles';

const _renderProductImage = (props) => (
  <View style={styles.imageViewContainer}>
    <Image
      style={styles.imageView}
      source={{uri: props.image}}
    />
  </View>
);

const _renderProductName = (props) => (
  <Text numberOfLines={1} style={styles.name}>{props.name}</Text>
);

const _renderProductPrice = (props) => (
  <View style={styles.priceContainer}>
    <Text style={styles.price}>{props.price}</Text>
    {props.off !== '' && <Text style={styles.off}>{props.off}</Text>}
  </View>
);

const _renderOutOfStock = () => (
  <View style={styles.outOfStockContainer}>
    <Text style={styles.outOfStock}>Out of stock</Text>
  </View>
);

const _renderNewTag = () => (
  <Text style={styles.new}>New</Text>
);

const _renderTags = (props) => {
  if (props.out_of_stock) return _renderOutOfStock()
  else if (props.new) return _renderNewTag();
  else return null
};

const _renderFavorite = (props) => (
  <View style={styles.favorite}>
    <Icon 
      size={20} 
      name={props.isFav ? "heart" : "heart-outline"} 
      color={props.isFav ? "#FE4F65" : "#898989"}
    />
  </View>
);

const _renderHalfStar = (index) => (
  <Icon 
    key={index}
    size={12} 
    name="star-half-sharp"
    color="#FECD12"
  />
);

const _renderFullStar = (index) => (
  <Icon
    key={index}
    size={12} 
    name="star-sharp"
    color="#FECD12"
  />
);

const _renderEmptyStar = (index) => (
  <Icon 
    key={index}
    size={12} 
    name="star-outline"
    color="#91918F"
  />
);

const _renderRating = (props) => {
  component = [];
  let i = 0;
  while (i < 5) {
    if (props.rating - i === 0.5) {
      component.push(_renderHalfStar(i))
    } else if (props.rating - i >= 1) {
      component.push(_renderFullStar(i))
    } else {
      component.push(_renderEmptyStar(i))
    }
    i++;
  }

  return (
    <View style={styles.rating}>
      {component}
    </View>
  );
};

const ProductItem = (props) => (
  <View style={styles.container}>
    {_renderTags(props)}
    {_renderFavorite(props)}
    {_renderProductImage(props)}
    {_renderRating(props)}
    {_renderProductName(props)}
    {_renderProductPrice(props)}
  </View>
);

export default ProductItem;