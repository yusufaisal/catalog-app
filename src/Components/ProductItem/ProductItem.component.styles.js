const shadow = {
  shadowColor: '#171717',
  shadowOffset: {width: 1, height: 1},
  shadowOpacity: 0.1,
  shadowRadius: 3,
}

export default {
  container: {
    flex: 1,
    backgroundColor: '#FFF',
    borderColor: '#ECECEC',
    borderWidth: 1,
    borderRadius: 10,
    paddingHorizontal: 12,
    paddingVertical: 12,
    marginHorizontal: 5,
    marginVertical: 5,
    ...shadow
  },
  imageViewContainer: {
    alignItems: 'center',
    justifyContent: 'center'
  },
  name: {
    fontWeight: '500',
    fontSize: 14,
    color: '#757479'
  },
  imageView: {
    width: 150,
    height: 150,
    resizeMode: 'contain'
  },
  priceContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginVertical: 5
  },
  price: {
    color: '#020202',
    fontSize: 18,
    fontWeight: '700',
  },
  off: {
    color: '#6D6B9C',
    fontWeight: '600',
  },
  outOfStockContainer: {
    position: 'absolute',
    left: 12,
    top: 12,
    alignSelf: 'flex-start',
    backgroundColor: '#FE4C4C',
    paddingVertical: 5,
    paddingHorizontal: 10,
    borderRadius: 10
  },
  outOfStock: {
    color: '#FFF',
    fontWeight: '600',
    fontSize: 12
  },
  new: {
    position: 'absolute',
    left: 12,
    top: 12,
    fontWeight: '500',
    fontSize: 16
  },
  favorite: {
    position: 'absolute',
    right: 12,
    top: 12,
  },
  rating: {
    flexDirection: 'row'
  }
};