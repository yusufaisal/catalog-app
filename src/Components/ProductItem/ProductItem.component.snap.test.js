import React from 'react';
import renderer from 'react-test-renderer';

import ProductItem from './ProductItem.component';

jest
  .mock('react-native-vector-icons/Ionicons', () => 'Icon');
  
describe('ProductItem Component', () => {
  const product = {
    "id": "1",
    "image": "https://fe.dev.dxtr.asia/storage/products/Product_07.png",
    "images": [
        "https://fe.dev.dxtr.asia/storage/products/Product_07.png",
        "https://fe.dev.dxtr.asia/storage/products/Product_02.png",
        "https://fe.dev.dxtr.asia/storage/products/Product_06.png"
    ],
    "name": "Pumma Runing Shoes",
    "price": "$350",
    "off": "20% Off",
    "rating": 4,
    "new": true,
    "isFav": true,
    "out_of_stock": false,
    "reviewCount": 150,
    "sizes": [
        6,
        7,
        8.5,
        32,
        36
    ],
    "description": "There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form, by injected humour, or randomised words which don't look even slightly believable. If you are going to use a passage of Lorem Ipsum, you need to be sure there isn't anything embarrassing hidden in the middle of text. All the Lorem Ipsum generators on the Internet tend to repeat predefined chunks as necessary, making this the first true generator on the Internet. It uses a dictionary of over 200 Latin words, combined with a handful of model sentence structures, to generate Lorem Ipsum which looks reasonable. The generated Lorem Ipsum is therefore always free from repetition, injected humour, or non-characteristic words etc."
  }

  it('should render ProductItem properly with new tag', () => {
    const tree = renderer
      .create(<ProductItem {...product}/>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render ProductItem properly with out of stock tag', () => {
    const tree = renderer
      .create(<ProductItem {...product} rating={3.5} out_of_stock={true}/>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render ProductItem properly without tag', () => {
    const tree = renderer
      .create(<ProductItem {...product} isFav={false} new={false}/>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});