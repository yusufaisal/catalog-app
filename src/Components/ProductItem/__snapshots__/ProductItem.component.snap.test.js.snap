// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem Component should render ProductItem properly with new tag 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderColor": "#ECECEC",
      "borderRadius": 10,
      "borderWidth": 1,
      "flex": 1,
      "marginHorizontal": 5,
      "marginVertical": 5,
      "paddingHorizontal": 12,
      "paddingVertical": 12,
      "shadowColor": "#171717",
      "shadowOffset": Object {
        "height": 1,
        "width": 1,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 3,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
        "fontWeight": "500",
        "left": 12,
        "position": "absolute",
        "top": 12,
      }
    }
  >
    New
  </Text>
  <View
    style={
      Object {
        "position": "absolute",
        "right": 12,
        "top": 12,
      }
    }
  >
    <Icon
      color="#FE4F65"
      name="heart"
      size={20}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://fe.dev.dxtr.asia/storage/products/Product_07.png",
        }
      }
      style={
        Object {
          "height": 150,
          "resizeMode": "contain",
          "width": 150,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#91918F"
      name="star-outline"
      size={12}
    />
  </View>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#757479",
        "fontSize": 14,
        "fontWeight": "500",
      }
    }
  >
    Pumma Runing Shoes
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#020202",
          "fontSize": 18,
          "fontWeight": "700",
        }
      }
    >
      $350
    </Text>
    <Text
      style={
        Object {
          "color": "#6D6B9C",
          "fontWeight": "600",
        }
      }
    >
      20% Off
    </Text>
  </View>
</View>
`;

exports[`ProductItem Component should render ProductItem properly with out of stock tag 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderColor": "#ECECEC",
      "borderRadius": 10,
      "borderWidth": 1,
      "flex": 1,
      "marginHorizontal": 5,
      "marginVertical": 5,
      "paddingHorizontal": 12,
      "paddingVertical": 12,
      "shadowColor": "#171717",
      "shadowOffset": Object {
        "height": 1,
        "width": 1,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "flex-start",
        "backgroundColor": "#FE4C4C",
        "borderRadius": 10,
        "left": 12,
        "paddingHorizontal": 10,
        "paddingVertical": 5,
        "position": "absolute",
        "top": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontSize": 12,
          "fontWeight": "600",
        }
      }
    >
      Out of stock
    </Text>
  </View>
  <View
    style={
      Object {
        "position": "absolute",
        "right": 12,
        "top": 12,
      }
    }
  >
    <Icon
      color="#FE4F65"
      name="heart"
      size={20}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://fe.dev.dxtr.asia/storage/products/Product_07.png",
        }
      }
      style={
        Object {
          "height": 150,
          "resizeMode": "contain",
          "width": 150,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-half-sharp"
      size={12}
    />
    <Icon
      color="#91918F"
      name="star-outline"
      size={12}
    />
  </View>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#757479",
        "fontSize": 14,
        "fontWeight": "500",
      }
    }
  >
    Pumma Runing Shoes
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#020202",
          "fontSize": 18,
          "fontWeight": "700",
        }
      }
    >
      $350
    </Text>
    <Text
      style={
        Object {
          "color": "#6D6B9C",
          "fontWeight": "600",
        }
      }
    >
      20% Off
    </Text>
  </View>
</View>
`;

exports[`ProductItem Component should render ProductItem properly without tag 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderColor": "#ECECEC",
      "borderRadius": 10,
      "borderWidth": 1,
      "flex": 1,
      "marginHorizontal": 5,
      "marginVertical": 5,
      "paddingHorizontal": 12,
      "paddingVertical": 12,
      "shadowColor": "#171717",
      "shadowOffset": Object {
        "height": 1,
        "width": 1,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 3,
    }
  }
>
  <View
    style={
      Object {
        "position": "absolute",
        "right": 12,
        "top": 12,
      }
    }
  >
    <Icon
      color="#898989"
      name="heart-outline"
      size={20}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://fe.dev.dxtr.asia/storage/products/Product_07.png",
        }
      }
      style={
        Object {
          "height": 150,
          "resizeMode": "contain",
          "width": 150,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#FECD12"
      name="star-sharp"
      size={12}
    />
    <Icon
      color="#91918F"
      name="star-outline"
      size={12}
    />
  </View>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#757479",
        "fontSize": 14,
        "fontWeight": "500",
      }
    }
  >
    Pumma Runing Shoes
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#020202",
          "fontSize": 18,
          "fontWeight": "700",
        }
      }
    >
      $350
    </Text>
    <Text
      style={
        Object {
          "color": "#6D6B9C",
          "fontWeight": "600",
        }
      }
    >
      20% Off
    </Text>
  </View>
</View>
`;
